internal/topology/check/compatibility.go

```diff
@@ -54,3 +62,261 @@ func ObjectsAreInTheSameNamespace(current, desired client.Object) error {
+					fmt.Sprintf("The %q MachineDeployment class can't be removed.", oldClass.Class),
+				),
+			)
+		}
+	}
+
+	// Ensure previous MachineDeployment class was modified in a compatible way.
+	for i, class := range in.Spec.Workers.MachineDeployments {
+		for _, oldClass := range old.Spec.Workers.MachineDeployments {
+			if class.Class == oldClass.Class {
+				// NOTE: class.Template.Metadata and class.Template.Bootstrap are allowed to change;
+				// class.Template.Bootstrap are ensured syntactically correct by validateAllRefs.
+				// Validates class.Template.Infrastructure template changes in a compatible way
+				if err := TemplatesAreCompatible(class.Template.Infrastructure, oldClass.Template.Infrastructure); err != nil {
+					allErrs = append(allErrs,
+						field.Invalid(field.NewPath("spec", "workers", "machineDeployments").Index(i).Child("template", "infrastructure"),
+							class.Template.Infrastructure.Ref,
+							err.Error(),
+						),
+					)
+				}
+			}
+		}
+	}
+	return allErrs
+}
+
+// MachineDeploymentClassesAreUnique checks that no two MachineDeploymentClasses in a ClusterClass share a name.
+func MachineDeploymentClassesAreUnique(in *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	classes := sets.NewString()
+	for i, class := range in.Spec.Workers.MachineDeployments {
+		if classes.Has(class.Class) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "workers", "machineDeployments").Index(i).Child("class"),
+					class.Class,
+					fmt.Sprintf("MachineDeployment class should be unique. MachineDeployment with class %q is defined more than once.", class.Class),
+				),
+			)
+		}
+		classes.Insert(class.Class)
+	}
+	return allErrs
+}
+
+// MachineDeploymentClassesAreUniqueAndDefinedInClusterClass checks that each MachineDeploymentClass is defined in the passed ClusterClass.
+func MachineDeploymentClassesAreUniqueAndDefinedInClusterClass(in *clusterv1.Cluster, class *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if in.Spec.Topology.Workers == nil {
+		return nil
+	}
+	// MachineDeployment class must be defined in the ClusterClass.
+	machineDeploymentClasses := classNamesFromWorkerClass(class.Spec.Workers)
+
+	names := sets.String{}
+	for _, md := range in.Spec.Topology.Workers.MachineDeployments {
+		if !machineDeploymentClasses.Has(md.Class) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "topology", "workers", "machineDeployments"),
+					md,
+					fmt.Sprintf("MachineDeployment class must be defined in the ClusterClass. MachineDeployment with class %q not found in ClusterClass %v.", md.Name, class.Name),
+				),
+			)
+		}
+		if names.Has(md.Name) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec", "topology", "workers", "machineDeployments"),
+					md,
+					fmt.Sprintf("MachineDeployment names should be unique. MachineDeployment with name %q is defined more than once.", md.Name),
```

Comment:
```suggestion
					fmt.Sprintf("MachineDeployment topology names should be unique. MachineDeployment with topology name %q is defined more than once.", md.Name),
```

Maybe that's better because those are not actually names of the MD. (additional context. those fields are exposed via `topologyName` via builtin variables.)